### ISSUE TITLE
"{ providedIn: 'root' }" does not instantiate services as documented

### ISSUE THREAD

**Issue by pfeileon** [angular/angular#25633]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Performance issue
[x] Feature request
[x] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ ] Other... Please describe:
</code></pre>

## Current behavior
If you provide a service by setting @Injectable({providedIn: 'root'}) the service is not created if it is never asked to be injected.

## Expected behavior
If you provide a service by setting {providedIn: 'root'} the service should be created according to the docs.
> When you provide the service at the root level, Angular creates a single, shared instance of HeroService and injects into any class that asks for it.

~~If this issue is actually the expected behavior,~~ (edit: it is) the docs should explicitly point out that the service is only instantiated if injected at least once. Also, I suggest adding a property for the @Injectable annotation to toggle this feature.

    // Example flag
    @Injectable({
        providedIn: 'root',
        eager: true // default: false
    })

## Minimal reproduction of the problem with instructions
Simply check your browser console and comment in/out the constructor in the app.module.ts file.

https://stackblitz.com/edit/angular-nsy3vj?file=src%2Fapp%2Fapp.component.ts

## Environment
`Angular version: 6.0.0`

**Comment by mlc-mlapis** [comment]
@pfeileon ... isn't it the reason why this new syntax was introduced? To have tree-shakable services?

**Comment by ericmartinezr** [comment]
I'm pretty sure this has been the behavior since like forever. The difference is that today the service is not part of your final bundle (tree-shaking), but services don't work by themselves, they need to be injected somewhere. It has always been like that.

If you add the `eager` property, what will you do with the service if it hasn't been injected anywhere?

**Comment by dariobraun** [comment]
What would be the preferred way to handle this? Add a disclaimer to the docs? Something like "Note that an instance of the Service is only created if there is at least one class asking to inject it."

**Comment by trotyl** [comment]
> Other types of providers can create their values lazily; that is, when they're needed for injection.

@dariobraun It's [already documented](https://angular.io/guide/dependency-injection-in-action#value-providers-usevalue), but there could be some docs change to make it more clear.

**Comment by mlc-mlapis** [comment]
@dariobraun ... but this is the basic assumption and the main reason why tree-shaking exits. If nothing is using such a service then why to have it in the code?

**Comment by dariobraun** [comment]
@mlc-mlapis Yeah, absolutely! To me it seems like this issue would not occur in any 'real' use cases and further clarification in the docs might not be necessary/useful.
If some final decision is made as to whether or not the docs should be updated, I'd be happy to work on a PR :)

**Comment by mlc-mlapis** [comment]
@dariobraun ... yep. Personally, I am used to repeating some facts because I want to be sure that it's absolutely understandable (better twice than once 😺 ) ... and it's my vice. So if you had any doubt then I vote for adding it into the docs.

**Comment by arunredhu** [comment]
> I'm pretty sure this has been the behavior since like forever. The difference is that today the service is not part of your final bundle (tree-shaking), but services don't work by themselves, they need to be injected somewhere. It has always been like that.
> 
> If you add the `eager` property, what will you do with the service if it hasn't been injected anywhere?

There might be use cases where you just need to add independent pluggable functionality that will not be used in any component or directive or service but you want a class which can use other services by injecting them and and do some calculations.

**Comment by HeeCheol-Kim** [comment]
I've followed the tutorial at https://angular.io/tutorial/toh-pt4 up, but it's not work properly.
MessageService doesn't work like totally unassigned instance. 

MessagesComponent, HeroService have a correct constructor for injectable service of MessageService.
but the line <div *ngIf="messageService.messages.Length">  doesn't react when I forced to add
some message string elements at MessageService's constructor.
![2019-02-02 205552](https://user-images.githubusercontent.com/41623476/52163939-0d337b00-272d-11e9-9851-496faa1c033e.jpg)
![2019-02-02 205609](https://user-images.githubusercontent.com/41623476/52163940-0d337b00-272d-11e9-9b67-a4b4e63e18a8.jpg)
![2019-02-02 205631](https://user-images.githubusercontent.com/41623476/52163941-0d337b00-272d-11e9-81dd-2b5b66fbc975.jpg)

**Comment by mlc-mlapis** [comment]
@HeeCheol-Kim ... and what is the `Length` in `<div *ngIf="messageService.messages.Length">`?

**Comment by HeeCheol-Kim** [comment]
> @HeeCheol-Kim ... and what is the `Length` in `<div *ngIf="messageService.messages.Length">`?

it means a member veriable(string array)'s count of MessageService class

**Comment by mlc-mlapis** [comment]
@HeeCheol-Kim ... really? Remember that JS is case sensitive.

**Comment by HeeCheol-Kim** [comment]
oh, I found it. the capital 'L' was the problem! Thanks. silly :)

**Comment by julianobrasil** [comment]
> If you add the eager property, what will you do with the service if it hasn't been injected anywhere?

@ericmartinezr, it's not the case that you haven't injected it anywhere. The main issue is that maybe you want it to be started along with the application to grab useful information that you will need in the future when you instantiate some component that injects it.

A classic example is when you want to keep track of the last route the user visited on the app (or maybe the user navigation history during the last hour). To do that, one way is to start a service that observes the navigation and store the last navigated url or something like that. And you just want to know about this navigation information, for example, in components where you have "BACK" buttons available to the user (not the browser ones, but your app provided buttons). You could build a service with some code like this: 

```ts
previousUrl = '';
currentUrl = '';

...

router.events.subscribe((event: RouterEvent) => {
      if (event instanceof NavigationEnd) {
        this.previousUrl = this.currentUrl;
        this.currentUrl = event.url;
      }
```

But if you don't inject the service in the `app.module.ts`, for example, it will not be instantiated until it's injected in some component that needs it... too late for it to be useful to that component. At that moment, if you need your navigation history to go back to the last visited page, you won't have it because the service would have just been started at that time.

**Comment by trotyl** [comment]
> it's not the case that you haven't injected it anywhere.

@julianobrasil If it's not injected, then likely no one has reference to that class, and it will be dropped away by tree-shaking without any trace at runtime. If you want additional bootstrap initialization logic, then [`APP_BOOTSTRAP_LISTENER`](https://angular.io/api/core/APP_BOOTSTRAP_LISTENER) is designed for that.

**Comment by rockument69** [comment]
To summarize: The section of the doc to be fixed is [Provide the HeroService](https://angular.io/tutorial/toh-pt4#provide-the-heroservice) in the [Tour of Heroes Pt. 4](https://angular.io/tutorial/toh-pt4). Key fixes are in [this comment](https://github.com/angular/angular/issues/25633#issuecomment-432152663) (proposed disclaimer) and [this comment](https://github.com/angular/angular/issues/25633#issuecomment-432157533) (existing documentation) to help understand the disclaimer. 

@trotyl , @matsko , @dariobraun , @mlc-mlapis , @ericmartinezr : If one of you can concur, I'll move forward with this issue to make the change. Thanks!

**Comment by Rahkola** [comment]
As a long-time user of various DI-containers, I also vouch for the idea of adding the eager: true flag for the @Injectable() decorator. I don't know if there are many other use-cases for eager instantiation but if the service only subscribes to events and is never injected, this feature would come in handy. 

I just found myself cleaning up the root level app.component.ts which is the place where devs (_including me_) put all the global event handling logic such as subscribing to the Angular router events. So being your average object-oriented Joe, I obviously ended up creating RouterEventHandlerService and isolating the event handling logic there. This approach both endorses single-responsibility-principle and solves the issue of polluting AppComponent with all sorts of event listeners.

Now this approach alone doesn't work of course since the service is not injected anywhere, but is supposed to just live as a singleton instance, instantiated at app start. While the workaround proposed [at Stack Overflow](https://stackoverflow.com/questions/41215226/make-sure-a-service-is-instantiated) works, I find it sub-optimal compared to the solution presented by OP.

**Comment by kemsky** [comment]
Well, certainly there are many cases when you want service to be instantiated immediately (eager): 

- pre-load something
- install listeners
-  do some global init

One could use `APP_INITIALIZER` for that, but in this case you have to export your services all the way to the top module, it's just not convenient, verbose and more importantly such code it is not self-evident:

```typescript
providers: [{
   provide: [APP_INITIALIZER],
   useFactory: (a, b, c) => Promise.resolve(),
   deps: [a, b, c],
   multi: true
  }]
```

**Comment by DmitryEfimenko** [comment]
based on the comment by @kemsky above, here's what I came up with:
```ts
export function emptyPromiseFactory() {
  return () => Promise.resolve();
}

export function eagerProvider(klass: Type<any>): Provider[] {
  return [
    klass,
    {
      provide: APP_INITIALIZER,
      useFactory: emptyPromiseFactory,
      deps: [klass],
      multi: true,
    },
  ];
}
```
In the AppModule:
```ts
providers: [
  eagerProvider(MyService),
]
```

**Comment by HarelM** [comment]
Just came across this issue when migrating my app to standalone components.
In my previous version, using a module, I injected a directive that will be used across the app for testing.
The components were not aware of this directive.
Now that they are standalone, I don't want to "litter" all my components with this directive, so I'm trying to create a service that will do a similar job to all the components (or something, I'm still not sure how to solve this TBH).
So I need this service initialized when the app starts, but I don't need it in any component or service so it's getting tree shaked...

Bottom line, I think standalone components might be a reason to reconsider adding this parameter.

**Comment by alxhub** [comment]
DI in Angular is lazy by design. _All_ services (not just `providedIn: 'root'`) are lazy and are only instantiated when requested.

`eager: true` is not an option we can support for `@Injectable`s. By definition, if nothing references the service, then the injector won't be aware of its existence (it won't even be included in the bundle) and can't instantiate it eagerly.

Generally, constructing services should not be side effectful. If you need a service to do something on application startup, the right pattern is to use an `ENVIRONMENT_INITIALIZER`:

```typescript
@Injectable()
class RandomLogger {
  // Note: no side effects in constructor

  initialize() {
    setInterval(() => console.log('...'), Math.round(Math.random() * 1000));
  }
}

function provideRandomLogger(): EnvironmentProviders {
  return makeEnvironmentProviders([
    RandomLogger,
    {provide: ENVIRONMENT_INITIALIZER, useValue: () => inject(RandomLogger).initialize(), multi: true},
  ]);
}
```
